File: src/main/java/work/xeltica/craft/core/models/MenuItem.java
package work.xeltica.craft.core.models;

import java.util.function.Consumer;

import org.bukkit.Material;
import org.bukkit.inventory.ItemStack;

/**
 * メニューのアイテム。
 * @author Xeltica
 */
public class MenuItem {
    public MenuItem(String name) {
        this(name, null);
    }

    public MenuItem(String name, Consumer<MenuItem> onClick) {
        this(name, onClick, Material.STONE_BUTTON);
    }

    public MenuItem(String name, Consumer<MenuItem> onClick, Material icon) {
       this(name, onClick, icon, null);
    }

    public MenuItem(String name, Consumer<MenuItem> onClick, Material icon, Object customData) {
        this(name, onClick, icon, customData, 1);
    }

    public MenuItem(String name, Consumer<MenuItem> onClick, Material icon, Object customData, boolean shiny) {
        this(name, onClick, icon, customData, 1, shiny);
    }

    public MenuItem(String name, Consumer<MenuItem> onClick, Material icon, Object customData, int count) {
        this(name, onClick, icon, customData, count, false);
    }

    public MenuItem(String name, Consumer<MenuItem> onClick, Material icon, Object customData, int count, boolean shiny) {
        this(name, onClick, new ItemStack(icon, count), customData, shiny);
    }

    public MenuItem(String name, Consumer<MenuItem> onClick, ItemStack icon) {
        this(name, onClick, icon, null);
    }

    public MenuItem(String name, Consumer<MenuItem> onClick, ItemStack icon, Object customData) {
        this(name, onClick, icon, customData, false);
    }

    public MenuItem(String name, Consumer<MenuItem> onClick, ItemStack icon, Object customData, boolean shiny) {
        this.name = name;
        this.onClick = onClick;
        this.icon = icon;
        this.customData = customData;
        this.shiny = shiny;
    }

    public String getName() { return name; }
    public Consumer<MenuItem> getOnClick() { return onClick; }
    public ItemStack getIcon() { return icon; }
    public Object getCustomData() { return customData; }
    public boolean isShiny() { return shiny; }

    private final String name;
    private final ItemStack icon;
    private final Consumer<MenuItem> onClick;
    private final Object customData;
    private final boolean shiny;
}

File: src/main/java/work/xeltica/craft/core/models/MenuItem.kt
package work.xeltica.craft.core.models

import org.bukkit.Material
import org.bukkit.inventory.ItemStack
import java.util.function.Consumer
import kotlin.jvm.JvmOverloads

/**
 * メニューのアイテム。
 * @author Xeltica
 */
class MenuItem @JvmOverloads constructor(
    val name: String?,
    val onClick: Consumer<MenuItem?>?,
    val icon: ItemStack,
    val customData: Any? = null,
    val isShiny: Boolean = false
) {
    constructor(name: String?, onClick: Consumer<MenuItem?>?, icon: Material?, customData: Any?, shiny: Boolean) : this(
        name,
        onClick,
        icon,
        customData,
        1,
        shiny
    ) {
    }

    @JvmOverloads
    constructor(
        name: String?,
        onClick: Consumer<MenuItem?>? = null,
        icon: Material? = Material.STONE_BUTTON,
        customData: Any? = null,
        count: Int = 1,
        shiny: Boolean = false
    ) : this(
        name, onClick, ItemStack(
            icon!!, count
        ), customData, shiny
    ) {
    }

}